content/00.build/65.developer-reference/50.era-contracts/60.elliptic-curve-precompiles.md

```diff
@@ -0,0 +1,311 @@
+we defer the divisions by multiplying them into a denominator.
+The denominator is represented by a new coordinate.
+Only at the very end, do we perform a single division to convert from projective coordinates back to affine coordinates.
+
+In affine form, each elliptic curve point has 2 coordinates, like $(x,y)$.
+In the new projective form, each point will have 3 coordinates, like $(X,Y,Z)$,
+with the restriction that $Z$ is never zero.
+The forward mapping is given by $(x,y)→(xz,yz,z)$, for any non-zero $z$ (usually chosen to be 1 for convenience).
+The reverse mapping is given by $(X,Y,Z)→(X/Z,Y/Z)$, as long as $Z$ is non-zero.
+
+### Point Doubling
+
+The affine form case $y=0$ corresponds to the projective form case $Y/Z=0$. This is equivalent to $Y=0$, since $Z≠0$.
+
+For the interesting case where $P=(X,Y,Z)$ and $Y≠0$, let’s convert the affine arithmetic to projective arithmetic.
+
+After expanding and simplifying the equations
+([demonstration here](https://www.nayuki.io/page/elliptic-curve-point-addition-in-projective-coordinates)),
+the following substitutions come out
+
+$$
+\begin{align*} T &= 3X^{2} + aZ^{2},\\ U &= 2YZ,\\ V &= 2UXY,\\ W &= T^{2} - 2V \end{align*}
+$$
+
+Using them, we can write
+
+$$
+\begin{align*} X_{r}  &= UW \\ Y_{r} &= T(V−W)−2(UY)^{2} \\ Z_{r} &= U^{3} \end{align*}
+$$
+
+As we can see, the complicated case involves approximately 18 multiplications, 4 additions/subtractions, and 0 divisions.
+
+### Point Addition
+
+The affine form case $x_{p} = x_{q}$ corresponds to the projective form case $X_{p}/Z_{p} = X_{q}/Z_{q}$.
+This is equivalent to $X_{p}Z_{q} = X_{q}Z_{p}$, via cross-multiplication.
+
+For the interesting case where $P = (X_{p},\ Y_{p},\ Z_{p})$ , $Q = (X_{q},\ Y_{q},\ Z_{q})$, and $X_{p}Z_{q} ≠ X_{q}Z_{p}$,
+let’s convert the affine arithmetic to projective arithmetic.
+
+After expanding and simplifying the equations
+([demonstration here](https://www.nayuki.io/page/elliptic-curve-point-addition-in-projective-coordinates)),
+the following substitutions come out
+
+$$
+\begin{align*}
+T_{0} &= Y_{p}Z_{q}\\
+T_{1} &= Y_{q}Z_{p}\\
+T &= T_{0} - T_{1}\\
+U_{0} &= X_{p}Z_{q}\\
+U_{1} &= X_{q}Z_{p}\\
+U &= U_{0} - U_{1}\\
+U_{2} &= U^{2}\\
+V &= Z_{p}Z_{q}\\
+W &= T^{2}V−U_{2}(U_{0}+U_{1}) \\
+\end{align*}
+$$
+
+Using them, we can write
+
+$$
+\begin{align*} X_{r}  &= UW \\ Y_{r} &= T(U_{0}U_{2}−W)−T_{0}U^{3} \\ Z_{r} &= U^{3}V \end{align*}
+$$
+
+As we can see, the complicated case involves approximately 15 multiplications, 6 additions/subtractions, and 0 divisions.
+
+## P256
+
+The P256Verify smart contract emulates the [RIP-7212](https://github.com/ethereum/RIPs/blob/master/RIPS/rip-7212.md)'s `P256VERIFY` precompile,
+enabling ECDSA signature verification using the secp256r1 curve (P-256) within the zkEVM environment.
+This contract acts as a wrapper for the associated circuit.
+It includes helper functions like unsafePackPrecompileParams to pack precompile parameters
```

Comment:
suggestion: Add code backtick for readability of function names
```suggestion
It includes helper functions like `unsafePackPrecompileParams` to pack precompile parameters
```

(same for `precompileCall` one line below)